packages/lake/src/hooks/usePersistedState.ts

```diff
@@ -1,26 +1,65 @@
 import { Option, Result } from "@swan-io/boxed";
-import { useCallback, useState } from "react";
+import { useCallback, useEffect, useState } from "react";
+
+const getItem = (key: string) => Result.fromExecution(() => localStorage.getItem(key)).getOr(null);
+
+const parseRawValue = <T>(rawValue: string | null, defaultValue: T) =>
+  Result.fromExecution(() => (rawValue != null ? (JSON.parse(rawValue) as T) : rawValue))
+    .toOption()
+    .flatMap(Option.fromNullable)
+    .getOr(defaultValue);
 
 export const usePersistedState = <T>(key: string, defaultValue: T) => {
-  const [state, setState] = useState<T>(() => {
-    return Result.fromExecution(() => localStorage.getItem(key))
-      .toOption()
-      .flatMap(Option.fromNullable)
-      .map(value => JSON.parse(value) as T)
-      .getOr(defaultValue);
+  const [state, setState] = useState(() => {
+    const rawValue = getItem(key);
+    const value = parseRawValue(rawValue, defaultValue);
+    return { defaultValue, value, rawValue };
   });
 
+  const updateRawValue = useCallback((rawValue: string | null) => {
+    setState(prevState => {
+      if (rawValue === prevState.rawValue) {
+        return prevState; // skip update if rawValue didn't changed
+      } else {
+        const { defaultValue } = prevState;
+        const value = parseRawValue(rawValue, defaultValue);
+        return { defaultValue, value, rawValue };
+      }
+    });
+  }, []);
+
   const setPersistedState = useCallback(
-    (state: T) => {
-      setState(state);
+    (value: T | null) => {
       try {
-        localStorage.setItem(key, JSON.stringify(state));
+        if (value != null) {
+          const rawValue = JSON.stringify(value);
+          updateRawValue(rawValue);
+          localStorage.setItem(key, rawValue);
+        } else {
+          updateRawValue(null);
+          localStorage.removeItem(key);
+        }
       } catch {
         // ignore
       }
     },
-    [key],
+    [key, updateRawValue],
   );
 
-  return [state, setPersistedState] as const;
+  useEffect(() => {
+    const listener = (event: StorageEvent) => {
+      if (event.key == null || event.key === key) {
```

Comment:
maybe add a check on `event.storageArea` being strictly equal to `localStorage`?

Reply:
Done